keps/sig-scheduling/3101-pod-to-node-assignment/README.md

```diff
@@ -0,0 +1,931 @@
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Customers trying to deploy batch workloads on Kubernetes have filed multiple
+requests (See [1](https://github.com/kubernetes/kubernetes/issues/68827),
+[2](https://github.com/kubernetes/kubernetes/issues/105287) for details) for 1:1
+pod-to-node assignment. The most common use case for enforcing 1:1 pod to node
+assignment is HPC workloads (MPI and Spark).
+
+Currently there are two approaches to achieve this:
+
+-   Set resource requests to something high enough to prevent scheduling any
+    other pods but daemonset pods. This is tricky to do and error prone,
+    especially for smaller nodes. This will not be possible if node size is not
+    known in advance.
+-   Use pod anti affinity. Users will need to define the same set of labels on
+    all pods that needs to be repelled, whether or not they actually request 1:1
+    pod to node assignment, it is also expensive to calculate especially for
+    cluster autoscaler.
+
+A simple and native way to force 1:1 pod-to-node assignment is therefore needed.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+-   Allow users to force 1:1 pod-to-node assignment for a given pod
+
+### Non-Goals
+
+-   Find the best fit node for a given pod
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Add a new pod-level API that allows users to specify how many "worker" pods are
+allowed on a node. The API is implemented by the scheduler as a filter plugin
+and kubelet as an admission predicate.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants a Pod to be scheduled alone on a node to avoid them from contending
+on resources (e.g. claiming all CPU cores available on a node). In other words,
+resource intensive workloads are needed to be scheduled on separate nodes.
+
+Example: As a user, I want to deploy a Spark or MPI job such that only a single
+worker is running on each pod excluding daemonsets.
```

Comment:
I can't tell from here, but it *sounds* like you want to run a Job that guarantees node exclusion (the items in this job run a number of pods that satisfy `min(nodesMatchingCriteria, numberOfReplicasRequested)`?

From an enhancement level perspective, I'm trying to understand the tradeoffs you need.  Could you add some of the properties you expect a solution to have in this section (like a list of criteria you expect to be true), and the use case that drives each property?  That helps me review, thank you!

Reply:
The use case is this: I want to run a single **workload** pod per node. One example is MPI or Spark: each MPI/Spark worker should be scheduled on a separate node alone; the only other pods allowed on the node are daemonsets (because they are used as node agents that serve a horizontal feature, like logging and monitoring)

Reply:
If there is not have availiable node for this workload pod, will this plugin evict other pod in some node?

Reply:
Yes if the incoming pod has higher priority

Reply:
OK, this should be clear in this KEP.

Reply:
Preemption applies to all filters, but doesn't hurt clarifying that in the doc.

Reply:
I added information about preemption in Design details section